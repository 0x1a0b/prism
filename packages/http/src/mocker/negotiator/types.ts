import { INodeExample, INodeExternalExample, ISchema } from '@stoplight/types';

export interface IHttpNegotiationResult {
  code: string;
  mediaType: string;
  example?: INodeExample | INodeExternalExample;
  schema?: ISchema;
}
