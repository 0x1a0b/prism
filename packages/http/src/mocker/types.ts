import { IOperationOptions, Protocol } from '@stoplight/prism-core/mocker/types';

export interface IHttpOperationOptions extends IOperationOptions {
  protocol: Protocol.HTTP;
  // Defaults to the lowest 2xx status code
  status: string;
  // Defaults to application/json
  mediaType: string;
  // Defaults to content type
  example?: string;
  // If true mock back response headers
  headers?: boolean;
}
