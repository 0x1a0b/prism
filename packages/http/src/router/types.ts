import { IHttpOperation } from '@stoplight/types/http';

export type Nullable<T> = T | null;

export interface IMatch {
  resource: IHttpOperation;
  serverMatch: MatchType | null;
  pathMatch: MatchType;
}

export enum MatchType {
  CONCRETE = 'concrete',
  TEMPLATED = 'templated',
  NOMATCH = 'no-match',
}
