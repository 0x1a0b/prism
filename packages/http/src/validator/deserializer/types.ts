import { ISchema } from '@stoplight/types/schema';

export type DeserializeHttpHeader = (value: string, type: string, explode: boolean) => any;

export interface IHttpHeaderParamStyleDeserializer {
  supports: (style: string) => boolean;
  deserialize: DeserializeHttpHeader;
}

export interface IHttpParamDeserializerRegistry<T extends Function> {
  get(style: string): T | undefined;
}

export interface IHttpParamStyleDeserializer<T> {
  supports: (style: string) => boolean;
  deserialize: T;
}

export type DeserializeHttpQuery = (
  key: string,
  query: {
    [name: string]: string | string[];
  },
  schema: ISchema,
  explode: boolean
) => any;

export interface IHttpQueryParamStyleDeserializer
  extends IHttpParamStyleDeserializer<DeserializeHttpQuery> {
  deserialize: DeserializeHttpQuery;
}
