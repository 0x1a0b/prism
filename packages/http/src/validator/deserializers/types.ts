import { HttpParamStyles, ISchema } from '@stoplight/types';

import { IHttpNameValue, IHttpNameValues } from '../../types';

export interface IHttpParamDeserializerRegistry<Parameters, S = HttpParamStyles> {
  get(style: S): IHttpParamStyleDeserializer<Parameters> | undefined;
}

export interface IHttpParamStyleDeserializer<Parameters, S = HttpParamStyles> {
  supports: (style: S) => boolean;
  deserialize: (name: string, parameters: Parameters, schema: ISchema, explode?: boolean) => any;
}

export type IHttpHeaderParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValue>;
export type IHttpQueryParamStyleDeserializer = IHttpParamStyleDeserializer<IHttpNameValues>;
