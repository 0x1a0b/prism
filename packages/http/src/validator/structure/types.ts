import { IValidation } from '@stoplight/prism-core';

export interface IHttpValidator<Obj, Specs> {
  validate(obj: Obj, specs: Specs[], mediaType?: string): IValidation[];
}
