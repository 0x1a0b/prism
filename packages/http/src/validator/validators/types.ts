import { IValidation } from '@stoplight/prism-core';
import { ISchema } from '@stoplight/types/schema';

export interface IHttpValidator<Target, Specs> {
  validate(target: Target, specs: Specs[], mediaType?: string): IValidation[];
}

export interface ISchemaValidator<S extends ISchema> {
  validate(content: any, schema: S): IValidation[];
  supports(mediaType: string): boolean;
}

export interface IValidatorRegistry {
  get(mediaType: string): ((content: any, schema: ISchema) => IValidation[]) | undefined;
}
