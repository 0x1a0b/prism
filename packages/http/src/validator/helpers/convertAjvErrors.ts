import { ValidationSeverity } from '@stoplight/prism-core/types';
import { ErrorObject } from 'ajv';

export function convertAjvErrors(
  errors: ErrorObject[],
  pathPrefix: string,
  severity: ValidationSeverity
) {
  return errors.map(error => ({
    path: [pathPrefix, ...error.dataPath.split('.').slice(1)],
    name: error.keyword || '',
    summary: error.message || '',
    message: error.message || '',
    severity,
  }));
}
