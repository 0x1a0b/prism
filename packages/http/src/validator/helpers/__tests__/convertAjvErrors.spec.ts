import { ValidationSeverity } from '@stoplight/prism-core/types';
import { convertAjvErrors } from '../convertAjvErrors';

describe('convertAjvErrors()', () => {
  const errorObjectFixture = {
    dataPath: 'a.b',
    keyword: 'required',
    message: 'c is required',
    schemaPath: '..',
    params: '',
  };

  describe('all fields defined', () => {
    it('converts properly', () => {
      expect(
        convertAjvErrors([errorObjectFixture], 'prefix', ValidationSeverity.ERROR)
      ).toMatchSnapshot();
    });
  });

  describe('keyword field is missing', () => {
    expect(
      convertAjvErrors(
        [Object.assign({}, errorObjectFixture, { keyword: undefined })],
        'prefix',
        ValidationSeverity.ERROR
      )
    ).toMatchSnapshot();
  });

  describe('message field is missing', () => {
    expect(
      convertAjvErrors(
        [Object.assign({}, errorObjectFixture, { message: undefined })],
        'prefix',
        ValidationSeverity.ERROR
      )
    ).toMatchSnapshot();
  });
});
