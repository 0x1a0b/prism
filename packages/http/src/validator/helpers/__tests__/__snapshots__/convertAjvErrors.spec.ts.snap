// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Object {
    "message": "c is required",
    "name": "",
    "path": Array [
      "prefix",
      "b",
    ],
    "severity": "error",
    "summary": "c is required",
  },
]
`;

exports[` 2`] = `
Array [
  Object {
    "message": "",
    "name": "required",
    "path": Array [
      "prefix",
      "b",
    ],
    "severity": "error",
    "summary": "",
  },
]
`;

exports[`convertAjvErrors() all fields defined converts properly 1`] = `
Array [
  Object {
    "message": "c is required",
    "name": "required",
    "path": Array [
      "prefix",
      "b",
    ],
    "severity": "error",
    "summary": "c is required",
  },
]
`;
